### PR TITLE
Various performance improvements

### LINKED ISSUE #865: Reduce texture copies, perhaps by uploading directly from ImageCesium to GPU?
Loading a texture in a tile involves the following copies:

1. Compressed image data comes over the network and stored in a glTF Buffer, base64 string, or separate image buffer loaded from a URL.
2. We decompress the image and store the raw pixels in `ImageCesium`. (in the case of KTX2, the pixel data isn't full decompressed)
3. The pixel data is copied into Unreal's `FTexturePlatformData`.
4. Unreal reportedly copies this _again_ on the game thread into a buffer owned by the render thread.
5. The buffer is uploaded to the GPU.

Steps 3 and 4, at least, seem unnecessary. There's no particular reason (to my knowledge) we can't keep the image in the ImageCesium alive as long as it's needed and upload it directly to the GPU. But it's unclear if it's possible to convince Unreal Engine to do such a thing.

### PR DESCRIPTION
**Pull request by nithinp7**
This PR contains several performance improvements, mostly focused on smoother frame rates during tile loads / tile destruction. 

_The original frame-times (in ms) throughout a fly-through the Melbourne tileset._
Baseline

_The frame-times during a similar fly-through, after the changes._
Current

In order of impact, the big changes are:

### **Threading Changes**
Moved async tasks to background workers. Previously async tile load tasks (draco decoding, creating physics meshes, etc.) executed on any worker thread they could find. Often, this meant that tile loads would saturate even foreground worker threads, preventing short but high-priority Unreal rendering tasks from executing. Any time there were more than a few simultaneous tile loads, this would _consistently_ cause Unreal's rendering thread to wait for a chance to execute asynchronous tasks. 

_Previously, worker threads were monopolized by tile loads. Notice the render thread dependency on a comparatively tiny task that gets stuck behind a huge tile load task - the render thread ends up waiting 50 ms in this case._
WorkerThreadMonopolization

Moving our async tile loads to background threads slows down tile loading speeds, due to fewer high-priority threads doing tile loads. From what I can tell we use the same number of worker threads after the change (10 that showed up in profiling on my machine) - but out of those 10 threads, we now seem to have 3 low-priority threads, that execute tasks much slower than the other 7 (again the breakdown is almost certainly machine-specific).

_This is how the Cesium async work currently looks in the profiler. Notice the three slow threads and 5 normal ones (there are 2 more normal ones not pictured)._
SlowThreads

(The statistics here have changed since the previously reported 66.7% increase)
In load-speed benchmarks using the Melbourne photogrammetry tileset inset with Cesium World Terrain + Bing Imagery, I noticed the following results for the same scene view:
- Before the change an initial scene load took an average of 3158ms.
- Currently in this branch, the initial load took on average 3862.75ms.
- This is a ~22.3% increase in load time. The variance between trials making up the above averages quite overshadows the difference in average load time, imho.

As an aside, a similar scene view in CesiumJS takes about 8800ms to load.

### **Destruction Changes**
Tweaked our destruction logic to avoid waiting for asynchronous cleanup tasks from finishing. Instead, we defer finalizing destruction of objects to a later frame if they are not immediately ready. During scenarios with high tile residency turnover (flying around a city close to the ground), render resources occasionally have to finish initializing before destruction. Previously, in such scenarios the game thread would end up essentially waiting for the render thread to finish initializing the tile so that we could begin destroying it.

_Previously, destruction tasks would keep queueing up more destruction tasks within the same tick, essentially making the game thread wait for async cleanup work._
OldDestruction

### **Texture Creation Changes**
Added two new code paths for texture resource creation:
Where supported, we can now create RHI textures asynchronously on worker threads as part of the tile load process. This saves a memcpy on the game thread and saves RHI texture creation on the render thread.

_The render thread cost of creating RHI textures before this PR. Notice the non-descript "Init Command" tasks on the render thread that are triggered by the game thread "UTexture2D::UpdateResource" - this is the RHI texture creation._
Ue5Trace1

~~This is almost always much better, but there are occasionally scenarios where multiple worker threads are creating RHI textures at the same time - this usually correlates to a GPU hitch in the profiler. I can't say if it _causes_ the GPU hitch though - it may be the case that when an unrelated GPU hitch happens it stalls any worker thread that tries to create a texture.~~ (this was due to a memory leak and no longer happens).

A second texture creation path has been added for when async texture creation is not supported. We now avoid the intermediary memcpy on the game thread by sending Unreal our in-memory gltf image buffer on the render thread. The image is then directly copied to the memory mapped GPU pointer. 

### **Main Thread Tile Load / Unload Throttling**
This is implemented in Cesium Native here: https://github.com/CesiumGS/cesium-native/pull/565

Previously, when tiles finished async load work, we _always_ executed their main thread load work that very frame. When tons of tiles finished loading at once, this would cause main thread hitches where we created several (sometimes 10+) tiles with all necessary UObjects (components, textures, etc) in a single tick. Now we have a budget for how much main thread load work we do every frame and it is prioritized in the same way as tile loads are. The default budget is 5ms.

Similarly, the time spent in unloadCachedTiles is now also throttled, with a default of 5ms per frame. This usually isn't a huge bottleneck, but this helps catch outliers.

### **Bonus: Unreal Insights + Cesium Native**
Added a hacky prototype for tracing Cesium Native with Unreal Insights. Not a user-facing feature, but extremely useful to see the bigger picture across all threads and across Unreal, Cesium for Unreal, and Cesium Native!

### **Follow-Up Steps for Performance**
- Take advantage of async texture creation for metadata textures (#979)
- Implement asynchronous vertex / index buffer creation (#982)
- Reduce / eliminate GC hitching (#980)
- Implement something similar to Unreal's DDC (derived data cache) for 3D Tiles (https://github.com/CesiumGS/cesium-native/issues/566)

A non-exhaustive list of TODOs before merging:
- [x] Clear up issue with CesiumGltf::ImageCesium pointer invalidation, especially for raster overlays.
- [x] Enable anisotropic filtering.
- [x] Make sure texture clamp mode is used (fix tile seams).
- [x] Fix encoded metadata textures, which are almost certainly broken.

Resolves #865

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @nithinp7!

* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] [Cesium for Unreal Samples](https://github.com/CesiumGS/cesium-unreal-samples) works.

File: Source/CesiumRuntime/Private/CesiumStaticMesh.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2020-2021 CesiumGS, Inc. and Contributors
+
+#include "CesiumStaticMesh.h"
+
+UCesiumStaticMesh::UCesiumStaticMesh(
+    const FObjectInitializer& ObjectInitializer)
+    : UStaticMesh(ObjectInitializer) {}
+
+bool UCesiumStaticMesh::IsReadyForFinishDestroy() {
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumStaticMesh.cpp:9]: Why is this needed, and why does it not need to call the base class implementation? Needs a comment at least.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 2]
```diff
@@ -1,18 +1,351 @@
...
+  uint32 GetSizeX() const override {
+    return static_cast<uint32>(_pCesiumImage->width);
+  }
+
+  uint32 GetSizeY() const override {
+    return static_cast<uint32>(_pCesiumImage->height);
+  }
+
+  virtual void InitRHI() override {
+    // TODO: does anisotropy work?
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:212]: I'm pretty sure it's not working. Raster overlays look very blurry on the horizon.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 3]
```diff
@@ -1,18 +1,351 @@
...
+    FSamplerStateInitializerRHI samplerStateInitializer(
+        SF_Trilinear,
+        AM_Wrap,
+        AM_Wrap,
+        AM_Wrap,
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:217]: Using wrap here instead of clamp is likely the cause of the seams I'm seeing between raster overlay tiles.
image

Comment (kring) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:217]: But wrap is the right answer in other scenarios, of course. It needs to be passed in from the sampler.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 4]
```diff
@@ -1,18 +1,351 @@
...
+  check(GRHISupportsAsyncTextureCreation);
+
+  ETextureCreateFlags textureFlags = TexCreate_ShaderResource | TexCreate_SRGB;
+
+  if (generateMipMaps) {
+    uint32 mipCount = static_cast<uint32>(image.mipPositions.size());
+
+    // TODO
+    // find way to reuse this heap allocation, static thread local??
+    std::vector<void*> mipsData(mipCount);
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:340]: mipCount will realistically never be larger than... what?... 16 maybe? So I'd just use a stack allocated array:

```
void* mipsData[16];
```

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 5]
```diff
@@ -1892,19 +1945,21 @@ static void loadPrimitiveGameThreadPart(
...
-      pbr,
-      pMaterial,
-      EMaterialParameterAssociation::GlobalParameter,
-      INDEX_NONE);
-  SetWaterParameterValues(
-      loadResult,
-      pMaterial,
-      EMaterialParameterAssociation::GlobalParameter,
-      INDEX_NONE);
+  // TODO: is this needed, caues applyTexture to run multiple times for same
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp]: Ah I see! Uncommented this code

File: Source/CesiumRuntime/Private/CesiumStaticMesh.cpp (hunk 1 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumStaticMesh.cpp]: Was an experiment that seemed to reduce destruction cost in UE5, but no longer needed after the destruction changes that prevent main thread blocking. Forgot to delete these files, removed them now!

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp (hunk 2 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: Should be fixed!

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp (hunk 3 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:255]: This should also be fixed now!

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp (hunk 4 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: Used your suggestion here, good idea!

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 6]
```diff
@@ -91,10 +92,32 @@ using TMeshVector4 = FVector4;
...
+    CesiumTextureUtility::destroyHalfLoadedTexture(*pHalfLoadedTexture.Get());
+  }
+}
 class HalfConstructedReal : public UCesiumGltfComponent::HalfConstructed {
 public:
-  virtual ~HalfConstructedReal() {}
   LoadModelResult loadModelResult;
+
+  virtual ~HalfConstructedReal() {
+    // TODO: deal with metadata case
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp]: Opened #979 and linked to it from the TODO. No special metadata cleanup needs to go there for now, since encoded metadata doesn't use async texture creation.

**Comment by nithinp7** [comment]
@kring Phew! I think this PR and the corresponding Cesium Native PR are ready for another review!

I did a quick smoke test for forbid holes and encoded metadata, and they both seem to be working so I'm happy. Just want to do one last smoke-test for Vulkan and maybe try it on mac before merging - but feel free to review meanwhile I feel good about the current state!

I updated the PR description here, so it should be accurate - I'll update the Cesium Native PR description in a second.

**Review by kring — COMMENTED** [review]
@nithinp7 this is amazing! So much smoother! A few things to address below, and also:

* [ ] Update CHANGES.md
* [x] Disable the tracing by default
* [ ] Make sure CI passes on all platforms (Failures might just be caused by the tracing)

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 7]
```diff
@@ -1,20 +1,429 @@
...
+    // mips.
+    uint32 mipCount = static_cast<uint32>(image.mipPositions.size());
+    if (mipCount > 16) {
+      mipCount = 16;
+    }
+
+    void* mipsData[16];
+    for (size_t i = 0; i < mipCount; ++i) {
+      const CesiumGltf::ImageCesiumMipPosition& mipPos = image.mipPositions[i];
+      mipsData[i] = (void*)&image.pixelData[mipPos.byteOffset];
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:387]: I just tried this, a static_cast seems to work in some of the instances where I was using the C-style cast. In other cases where Unreal's API expected a void* for _reading only_, and we had a const reference to buffer or some such - static_cast<void*> isn't allowed since it attempts to cast away the const qualifier. In that case should reinterpret_cast be used? Or const_cast, somehow?

File: extern/CMakeLists.txt [hunk 8]
```diff
@@ -68,6 +68,25 @@ elseif (${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zp8")
 endif()
 
+
+# Trace Cesium Native with Unreal Insights. Need to include Unreal Engine headers when building 
+# Cesium Native.
+
+# Change this line to change engine version. Note you will have to set an environment variable to
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply extern/CMakeLists.txt:75]: Sweet I was looking for a way to set that up!

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 9]
```diff
@@ -1918,12 +1961,16 @@ void ACesium3DTileset::Tick(float DeltaTime) {
 
   showTilesToRender(result.tilesToRenderThisFrame);
 
-  for (Cesium3DTilesSelection::Tile* pTile : result.tilesToRenderThisFrame) {
-    updateTileFade(pTile, true);
-  }
+  if (this->UseLodTransitions) {
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/Cesium3DTileset.cpp:1964]: Ah yup good catch, we could probably call updateTileFade for all loaded tiles when the UseLodTransitions option is changed - at least that would be better than triggering an entire tileset reload.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 10]
```diff
@@ -1,20 +1,429 @@
...
+  uint32 GetSizeY() const override { return this->_height; }
+
+  virtual void InitRHI() override {
+    FSamplerStateInitializerRHI samplerStateInitializer(
+        this->_filter,
+        this->_addressX,
+        this->_addressY,
+        AM_Wrap,
+        0.0f,
+        4,
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: Passing in 0 seems to do the trick!

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/Cesium3DTileset.cpp:1971]: This should be fixed now.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 11]
```diff
@@ -1,20 +1,429 @@
...
+      }
+
+      this->TextureRHI = rhiTexture;
+      rhiTexture.SafeRelease();
+    }
+
+    RHIUpdateTextureReference(TextureReferenceRHI, this->TextureRHI);
+
+    // TODO: investigate if we can / should do this!
+    // related to: https://github.com/CesiumGS/cesium-unreal/issues/980
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: I investigated this again, and it doesn't look to make much of a difference. Deferred delete just means it waits for a render thread - rhi thread sync so it can safely delete the rhi object (from what I can tell). Explicitly not deferring delete might a) be unsafe, or b) force an unnecessary sync-up.

I can't be too confident about my current understanding of this, but I'm almost positive it's not related to the GC pressure we're seeing now - so I removed the comment! (There's no other issue that we could or need to solve by using DoNoDeferDelete anyways)

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 12]
```diff
@@ -110,137 +574,42 @@ TUniquePtr<LoadedTextureResult> loadTextureAnyThreadPart(
...
+  // Mark the image source as legacy, so we later know where to look for image
+  // data.
+  pResult->textureSource = LegacyTextureSource{};
+#else
+  if (GRHISupportsAsyncTextureCreation) {
+    // Create RHI texture resource asynchronously.
+    std::string scopeName =
+        "Cesium::CreateRHITexture2D" + std::to_string(image.width) + "x" +
+        std::to_string(image.height) + "x" + std::to_string(image.channels) +
+        "x" + std::to_string(image.bytesPerChannel);
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: I just removed the dynamic string from the tracing - it wasn't super useful anyways, the size of the textures (within reason) didn't cause bottlenecks. If we really need to know that type of information in the future it should probably go into some sort of Unreal stats system, not squeezed into the trace name!

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 13]
```diff
@@ -1,20 +1,429 @@
...
+
+    RHIUpdateTextureReference(TextureReferenceRHI, this->TextureRHI);
+
+    // TODO: investigate if we can / should do this!
+    // related to: https://github.com/CesiumGS/cesium-unreal/issues/980
+    // this->TextureRHI->DoNoDeferDelete();
+    // this->TextureReferenceRHI->DoNoDeferDelete();
+  }
+
+  virtual void ReleaseRHI() override {
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:334]: I think you're right about the TextureRHI.SafeRelease call here, I removed it with no issues. I'll admit I spent a large amount of time switching things around and trying nonsensical things in this function trying to fix a completely unrelated memory leak. 

The TextureReferenceRHI.SafeRelease call I'm not super sure about, I can't find anywhere in the engine that this is explicitly released (it probably just gets implicitly released when the Resource object is deleted). So I'm going to remove that too, it doesn't seem to cause any issues from what I can tell.

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 14]
```diff
@@ -1,20 +1,429 @@
...
+ * @param generateMipMaps Whether the RHI texture should have a mipmap.
+ * @return The RHI texture reference.
+ */
+FTexture2DRHIRef CreateRHITexture2D_Async(
+    const CesiumGltf::ImageCesium& image,
+    EPixelFormat format,
+    bool generateMipMaps) {
+  check(GRHISupportsAsyncTextureCreation);
+
+  ETextureCreateFlags textureFlags = TexCreate_ShaderResource | TexCreate_SRGB;
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: Should be fixed now

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp [hunk 15]
```diff
@@ -1,20 +1,429 @@
...
+        this->_filter,
+        this->_addressX,
+        this->_addressY,
+        AM_Wrap,
+        0.0f,
+        1,
+        0.0f,
+        2.0f);
+    this->DeferredPassSamplerStateRHI =
+        GetOrCreateSamplerState(deferredSamplerStateInitializer);
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:288]: They were taken from here:
https://github.com/EpicGames/UnrealEngine/blob/d9d435c9c280b99a6c679b517adedd3f4b02cfd7/Engine/Source/Runtime/Engine/Private/Rendering/StreamableTextureResource.cpp#L250-L264

The runtime texture loading plugin also had copied these same exact lines. I added the comments from the above link to our code to give context (I don't understand it perfectly either).

**Comment by nithinp7** [comment]
@kring I hit resolve on the comments I think I addressed and checked off all the TODOs except for CI. There's just the one comment remaining about different types of casting, let me know what you think there. But otherwise I think I'm happy with this! I'm signing off for tonight but hopefully CI will pass - it just looked like a minor API difference in 4.26 that I think should be fixed now. Thanks for reviewing!

File: Source/CesiumRuntime/Private/CesiumTextureUtility.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kring) [reply Source/CesiumRuntime/Private/CesiumTextureUtility.cpp]: Ah ok. The C++ purist way would be to use two casts (static_cast to `const void*` followed by const_cast to `void*`), but the C-style is fine.

**Comment by kring** [comment]
Great work @nithinp7, merging!
